### PR TITLE
Fix escaped markdown passing backslashes through

### PR DESCRIPTION
**Pull request by BobVul**
Fixes https://github.com/vector-im/riot-web/issues/11230

This implements the simpler approach I suggested in the linked issue, namely always returning the CommonMark parsed result even if we think it's plaintext. This is because the current plaintext check only checks after any backslash escapes were processed by CommonMark, but if we then send the pre-parse message that results in the backslashes becoming visible in the message.

This currently causes test failures in tests (currently 3) that do not expect any HTML content, since it causes HTML content to be *always* generated and sent unless `/plain` is explicitly requested.

If this absolutely cannot happen (and we can't change those tests accordingly), then an alternative is to only send the post-parse 'plaintext' if it differs from the source markdown, e.g.:

```js
    const postParsePlaintext = parser.toPlaintext();
    if (postParsePlaintext !== md) {
        return postParsePlaintext;
    }
```

Not sure which approach (change tests vs only send when different) is better.

### CODE REVIEW

**Comment by BobVul** [comment]
I've tentatively gone with only returning formatted text where it differs from source text. This passes the existing tests that expect plain text.

Still not sure if it's the best approach (always returning formatted text is more consistent). It's not worse than the current state at least.

**Comment by BobVul** [comment]
Is there anything I need to do to progress this PR?

**Comment by turt2live** [comment]
ah, sorry. Looks like it got lost in the depths of my notifications. I'll send it through CI and get it into the team's review queue (as it might not be Bruno who reviews this, though there is a chance).

**Comment by BobVul** [comment]
Sorry, I think I just clicked "request review" and it suggested someone for me.

Thanks for looking at this.

**Comment by BobVul** [comment]
Thanks for the review.

> I think we should preserve the behaviour of only setting `formatted_body` on a message when it is needed. Correct me if I am wrong, but I think that as it stands, every message will be sent with both a `body` and `formatted_body`.

The current commit only sends a `formatted_body` if it differs from the `body`. Specifically, if the output of `toPlaintext` is different from the source `md`. It does not return `formatted_body` for inputs not containing any Markdown characters at all (e.g. `'hello world'`). I'm not sure this is a *perfect* solution, but my theory here is that if the Markdown parser output differs from the input then there must have been some Markdown (including escape sequences) that it had to process.

Please note that it does pass existing tests that verify this behaviour of only containing a `body` when no formatting is requested at all.

The only caveat is that `body` will contain the original source text (`"\*hello world\*"`) while `formatted_body` contains the parsed/rendered output (`"*hello world*"`), but I view that as expected behaviour. It is consistent with other messages that contain a mix of escaped and unescaped Markdown.

The way I see it, there are four input cases:

1. Only has active markdown
2. Only has escaped markdown
3. Has both active and escaped markdown
4. Has no markdown

I think cases 1 through 3 should be treated the same way. Only case 4 should be an exception, and this is what the current PR accomplishes.

---

Current PR:

| Case | Input | `body` | `formatted_body` |
| --- | --- | --- | --- |
| 1 | `"*test*"` | `"*test*"` | `"<em>test</em>"` |
| 2 | `"\*test\*"` | `"\*test\*"` | `"*test*"` |
| 3 | `"\*foo\* *bar*"` | `"\*foo\* *bar*"` | `"*foo* <em>bar</em>"` |
| 4 | `"test"` | `"test"` | `undefined` |

---

> After looking into it a bit, I thought that returning `text + new Markdown(part.text).toPlaintext()` rather than `part.text` for a `plain` part from `textSerialize` would be the way to go here to preserve the above behaviour, but apparently `toPlaintext` can return HTML. E.g. `new Markdown("__foo__").toPlaintext()` returns `"<strong>foo</strong>"`. So that's not ideal either. `toPlaintext` is implemented on top of the commonmark parser, and I probably won't have the bandwidth to look into how to fix it to not generate HTML and only do backslash conversion before we have cross-signing is released.

Changing `textSerialize` was the other option I had considered (in the linked issue), and ran into the same problem: there is no easy way to *only* unescape the input without generating HTML. CommonMark 0.29.1 does expose a base `Renderer` class but it does not appear to actually do anything meaningful. If this approach is required, we'd need to write a custom renderer (which wouldn't necessarily do anything complex, but would at least need to implement all the possible types).

I also feel this way is more inconsistent, since we would be treating two conceptually similar cases very differently. Going back to the cases above, this would be treating cases 1 and 3 the same, and 2 and 4 the same. But I do not see why 2 should be different from 1 or 3.

---

This would look something like:

| Case | Input | `body` | `formatted_body` |
| --- | --- | --- | --- |
| 1 | `"*test*"` | `"*test*"` | `"<em>test</em>"` |
| 2 | `"\*test\*"` | `"*test*"` | `undefined` |
| 3 | `"\*foo\* *bar*"` | `"\*foo\* *bar*"` | `"*foo* <em>bar</em>"` |
| 4 | `"test"` | `"test"` | `undefined` |

---

> So if you'd like to keep working on this, I would suggest to continue with the approach to adapt `textSerialize` and see if there is a way to fix `toPlaintext`.

I do favour the current approach, given it only returns `formatted_body` when it differs (including escaped markdown) from `body`. If you still think it is necessary to change `textSerialize` instead, I can look into creating a custom Markdown renderer - but I'm not sure if that's actually any better from consistency, effort, or maintenance perspectives.

**Comment by bwindels** [comment]
Thank you for the extensive documentation of the issue, that helps to clarify things.

> The current PR since the second commit (9cf53aa) only sends a formatted_body if it differs from the body.

Indeed, I missed that point. I do still think though that the ideal solution would be to treat case 2 and 4 from your table as plain text, and only send case 1 and 3 with `formatted_html`.

The spec isn't super clear on this, but my reasoning is that `body` is supposed to be a plain text representation of the message, not the literal input of the user into a text field. The fact that we support backslashes is a consequence of Riot supporting markdown, which is unrelated to the protocol. `body` is meant to be read by clients that don't support markup, and putting input artefacts like backslashes in there is not optimal. Furthermore, a message, as the user intended it, with just backslashes can be represented without any loss of information with just plain text (by removing them). `formatted_html` on the other hand, is only to be used when any markup is needed.

E.g. to send `formatted_html` or not is not about whether the input contains things that would be transformed by a markdown parser, it's rather about the fact that a message needs markup to be shown without loss of information.

Having said that, this PR already brings an improvement to the current situation of just not supporting backslashes. It still breaks for clients that prefer to just use `body`, and is more chatty than it needs to be, but right now, it is broken for all clients.

So, in short, I would prefer the solution with a custom text renderer (where you can probably also implement the "needs markup?" check more reliably while rendering as text) but I don't have the time to do the work right now, so depending on your preference, I can merge this as it is.


In the mean time, I'll check whether this PR passed all build checks.

**Comment by BobVul** [comment]
> I do still think though that the ideal solution would be to treat case 2 and 4 from your table as plain text, and only send case 1 and 3 with `formatted_html`.
> 
> The spec isn't super clear on this, but my reasoning is that `body` is supposed to be a plain text representation of the message, not the literal input of the user into a text field. The fact that we support backslashes is a consequence of Riot supporting markdown, which is unrelated to the protocol. `body` is meant to be read by clients that don't support markup, and putting input artefacts like backslashes in there is not optimal. Furthermore, a message, as the user intended it, with just backslashes can be represented without any loss of information with just plain text (by removing them). `formatted_html` on the other hand, is only to be used when any markup is needed.
> 
> E.g. to send `formatted_html` or not is not about whether the input contains things that would be transformed by a markdown parser, it's rather about the fact that a message needs markup to be shown without loss of information.

That is a good point. Admittedly I've not really done much more than skim the spec briefly, and it did not say much about `formatted_body` apart from saying it's meant to contain HTML.

I still feel there's going to be some confusion for plain-text clients who then can't tell the difference between `*foo*` and `\*foo\*` but I suppose there's not much that can be done to solve that issue.

> So, in short, I would prefer the solution with a custom text renderer (where you can probably also implement the "needs markup?" check more reliably while rendering as text) but I don't have the time to do the work right now, so depending on your preference, I can merge this as it is.

I think I can clobber a basic literal/plaintext renderer together, though that may have to wait until I have a bit more time next week. Would it be better as a separate PR? It would not share any code with this one.

A proper plaintext renderer would would do more or less the same thing the current "needs markdown" check does (step through all nodes checking if there are non-text ones).

**Comment by bwindels** [comment]
> I still feel there's going to be some confusion for plain-text clients who then can't tell the difference between `*foo*` and `\*foo\*` but I suppose there's not much that can be done to solve that issue.

There should be no confusion?

When entering `*foo*`, `body` should contain `foo` and loose the markup. This is the trade-off that clients make that don't support `formatted_html`.

When entering `\*foo\*`, the user actually meant to enter `*foo*` as plain text, so that is what `body` should contain. There should be no `formatted_body` as there is no markup.

It's worth noting that other clients might be completely unaware of markdown. The only format present in the spec is (a subset of) HTML.

> I think I can clobber a basic literal/plaintext renderer together, though that may have to wait until I have a bit more time next week. Would it be better as a separate PR? It would not share any code with this one.

Yeah, a separate PR sounds great, thanks! I'll have a look on Monday whether I want to merge this in the meantime, as it's getting late now.

**Comment by bwindels** [comment]
Filed https://github.com/vector-im/riot-web/issues/12371 to track the follow-up work.

**Comment by BobVul** [comment]
Thanks! I do still intend to look at the follow-up; unfortunately I didn't had enough time this week.

**Comment by bwindels** [comment]
I reverted this because it doesn't escape < in plain text message, which are posted with `formatted_html` with these changes.

E.g. the message "looks like we're not escaping '<' ☹️" (with the emoji being auto-completed to from the :( emoticon encoding) is sent as
```
"formatted_body": "looks like we're not escaping '<' ☹️"
```
whereas really be sent as just a `body`, but if sent as a `formatted_body` should be sent as:
```
"formatted_body": "looks like we're not escaping '&lt;' ☹️"
```
